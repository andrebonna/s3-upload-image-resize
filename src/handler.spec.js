import fs from 'fs';
import AWS from 'aws-sdk';

function generatePromiseFn(content) {
    return jest.fn(() => ({ promise: () => Promise.resolve(content) }));
}

describe('S3 Upload', () => {

    const body = fs.readFileSync(`${__dirname}/mockImage.jpg`);
    let getObjectSpy = generatePromiseFn({ Body: body });
    let putObjectSpy = generatePromiseFn();
    let deleteObjectSpy = generatePromiseFn();
    let putSpy = generatePromiseFn();
    let deleteSpy = generatePromiseFn();
    AWS.S3 = class {
        constructor() {
            this.getObject = getObjectSpy;
            this.putObject = putObjectSpy;
            this.deleteObject = deleteObjectSpy;
        }
    };
    AWS.DynamoDB.DocumentClient = class {
        constructor() {
            this.put = putSpy;
            this.delete = deleteSpy;
        }
    };
    const { create, remove } = require('./handler');

    beforeEach(() => {
        getObjectSpy.mockClear();
        putObjectSpy.mockClear();
        deleteObjectSpy.mockClear();
        putSpy.mockClear();
        deleteSpy.mockClear();
    });

    it('create handler successfully', (done) => {
        create({
            Records: [{
                s3: {
                    object: {
                        key: 'keyMocked'
                    }
                }
            }]
        }).then(() => {
            expect(getObjectSpy.mock.calls).toMatchSnapshot();
            expect(putObjectSpy.mock.calls.length).toBe(3);
            expect(putSpy.mock.calls).toMatchSnapshot();
            done();
        }).catch(done.fail);
    });

    it('create handler with S3 getObject error', (done) => {

        getObjectSpy.mockImplementationOnce(() => ({
            promise: () => Promise.reject('simulate getObject error')
        }));
        create({
            Records: [{
                s3: {
                    object: {
                        key: 'keyMocked'
                    }
                }
            }]
        }).then(() => {
            expect(getObjectSpy.mock.calls).toMatchSnapshot();
            expect(putObjectSpy.mock.calls.length).toBe(0);
            expect(putSpy.mock.calls).toMatchSnapshot();
            done();
        }).catch(done.fail);
    });

    it('create handler with S3 putObject error', (done) => {

        putObjectSpy.mockImplementationOnce(() => ({
            promise: () => Promise.reject('simulate putObject error')
        }));
        create({
            Records: [{
                s3: {
                    object: {
                        key: 'keyMocked'
                    }
                }
            }]
        }).then(() => {
            expect(getObjectSpy.mock.calls).toMatchSnapshot();
            expect(putObjectSpy.mock.calls.length).toBe(1);
            expect(putSpy.mock.calls).toMatchSnapshot();
            done();
        }).catch(done.fail);
    });

    it('create handler with DynamoDB put error', (done) => {

        putSpy.mockImplementationOnce(() => ({
            promise: () => Promise.reject('simulate DynamoDB put error')
        }));
        create({
            Records: [{
                s3: {
                    object: {
                        key: 'keyMocked'
                    }
                }
            }]
        }).then(() => {
            expect(getObjectSpy.mock.calls).toMatchSnapshot();
            expect(putObjectSpy.mock.calls.length).toBe(3);
            expect(putSpy.mock.calls).toMatchSnapshot();
            done();
        }).catch(done.fail);
    });

    it('remove handler successfully', (done) => {
        remove({
            Records: [{
                s3: {
                    object: {
                        key: 'keyMocked'
                    }
                }
            }]
        }).then(() => {
            expect(deleteObjectSpy.mock.calls).toMatchSnapshot();
            expect(deleteSpy.mock.calls).toMatchSnapshot();
            done();
        }).catch(done.fail);
    });

    it('remove handler with S3 deleteObject error', (done) => {
        deleteObjectSpy.mockImplementationOnce(() => ({
            promise: () => Promise.reject('simulate deleteObject error')
        }));
        remove({
            Records: [{
                s3: {
                    object: {
                        key: 'keyMocked'
                    }
                }
            }]
        }).then(() => {
            expect(deleteObjectSpy.mock.calls).toMatchSnapshot();
            expect(deleteSpy.mock.calls).toMatchSnapshot();
            done();
        }).catch(done.fail);
    });

    it('remove handler with DynamoDB delete error', (done) => {
        deleteSpy.mockImplementationOnce(() => ({
            promise: () => Promise.reject('simulate DynamoDB delete error')
        }));
        remove({
            Records: [{
                s3: {
                    object: {
                        key: 'keyMocked'
                    }
                }
            }]
        }).then(() => {
            expect(deleteObjectSpy.mock.calls).toMatchSnapshot();
            expect(deleteSpy.mock.calls).toMatchSnapshot();
            done();
        }).catch(done.fail);
    });
});