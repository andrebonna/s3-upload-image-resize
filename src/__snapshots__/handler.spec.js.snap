// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`S3 Upload create handler successfully 1`] = `
Array [
  Array [
    Object {
      "Bucket": undefined,
      "Key": "keyMocked",
    },
  ],
]
`;

exports[`S3 Upload create handler successfully 2`] = `
Array [
  Array [
    Object {
      "Item": Object {
        "Categories": Array [
          "nature",
          "urban",
        ],
        "Description": "bisse le bissè",
        "Name": "keyMocked",
        "Title": "Teste malucão de título",
        "URLs": Array [
          "https://s3.amazonaws.com/undefined/keyMocked-1024.png",
          "https://s3.amazonaws.com/undefined/keyMocked-640.png",
          "https://s3.amazonaws.com/undefined/keyMocked-320.png",
        ],
      },
      "TableName": undefined,
    },
  ],
]
`;

exports[`S3 Upload create handler with DynamoDB put error 1`] = `
Array [
  Array [
    Object {
      "Bucket": undefined,
      "Key": "keyMocked",
    },
  ],
]
`;

exports[`S3 Upload create handler with DynamoDB put error 2`] = `
Array [
  Array [
    Object {
      "Item": Object {
        "Categories": Array [
          "nature",
          "urban",
        ],
        "Description": "bisse le bissè",
        "Name": "keyMocked",
        "Title": "Teste malucão de título",
        "URLs": Array [
          "https://s3.amazonaws.com/undefined/keyMocked-1024.png",
          "https://s3.amazonaws.com/undefined/keyMocked-640.png",
          "https://s3.amazonaws.com/undefined/keyMocked-320.png",
        ],
      },
      "TableName": undefined,
    },
  ],
]
`;

exports[`S3 Upload create handler with S3 getObject error 1`] = `
Array [
  Array [
    Object {
      "Bucket": undefined,
      "Key": "keyMocked",
    },
  ],
]
`;

exports[`S3 Upload create handler with S3 getObject error 2`] = `Array []`;

exports[`S3 Upload create handler with S3 putObject error 1`] = `
Array [
  Array [
    Object {
      "Bucket": undefined,
      "Key": "keyMocked",
    },
  ],
]
`;

exports[`S3 Upload create handler with S3 putObject error 2`] = `Array []`;

exports[`S3 Upload remove handler successfully 1`] = `
Array [
  Array [
    Object {
      "Bucket": undefined,
      "Key": "keyMocked-1024.png",
    },
  ],
  Array [
    Object {
      "Bucket": undefined,
      "Key": "keyMocked-640.png",
    },
  ],
  Array [
    Object {
      "Bucket": undefined,
      "Key": "keyMocked-320.png",
    },
  ],
]
`;

exports[`S3 Upload remove handler successfully 2`] = `
Array [
  Array [
    Object {
      "Key": Object {
        "Name": "keyMocked",
      },
      "TableName": undefined,
    },
  ],
]
`;

exports[`S3 Upload remove handler with DynamoDB delete error 1`] = `
Array [
  Array [
    Object {
      "Bucket": undefined,
      "Key": "keyMocked-1024.png",
    },
  ],
  Array [
    Object {
      "Bucket": undefined,
      "Key": "keyMocked-640.png",
    },
  ],
  Array [
    Object {
      "Bucket": undefined,
      "Key": "keyMocked-320.png",
    },
  ],
]
`;

exports[`S3 Upload remove handler with DynamoDB delete error 2`] = `
Array [
  Array [
    Object {
      "Key": Object {
        "Name": "keyMocked",
      },
      "TableName": undefined,
    },
  ],
]
`;

exports[`S3 Upload remove handler with S3 deleteObject error 1`] = `
Array [
  Array [
    Object {
      "Bucket": undefined,
      "Key": "keyMocked-1024.png",
    },
  ],
]
`;

exports[`S3 Upload remove handler with S3 deleteObject error 2`] = `Array []`;
